### DISCUSSION TITLE
Enforcing auth on custom endpoints

### DISCUSSION THREAD

**Discussion by ntatko** [directus/directus#5300] [Archive]
So, I have a custom endpoint - `POST` to `custom/courses`. I'm sending along the `Authorization` header with a proper authenticated user's token (I can post to `/items/lists` with the token, which is a restricted endpoint), but I don't need to, there's no logic to mandate that the auth token is respected. 

Are there docs around securing custom endpoints, or do I need to do that programmatically? Does anyone have any ideas on where to find some docs or other questions about this?

Thanks for any help you might have!

**Comment by rijkvanzanten** [comment]
The docs are limited on this (cc @nickrum), but the custom endpoints example does include an authorized call to the ItemsService:

https://docs.directus.io/guides/api-endpoints/#full-example

The request contains information on the current user in `req.accountability`. This object can be passed on to `ItemsService` (or any of the other services) to kick in permissions checks etc.

Another usage example would be the actual Directus API itself:

https://github.com/directus/directus/blob/207eda1d2dd35a6d96721f32c511a01bb6e37030/api/src/controllers/items.ts#L94-L117

Note how `req.accountability` is passed to `ItemsService`

**Reply by benosman** [reply]
@rijkvanzanten 

This code snippet was helpful. Thank you.

On a similar note, is there a way to enable caching for custom endpoints? 

Looking at the endpoints in the Directus API, including the one you included above, it looks like the respond middleware is reponsible for setting the cache. Is there a way to import/require that middleware into an endpoint extension?

**Update:**

I worked it out, I can use the following.
```js
const { respond } = require('directus/dist/middleware/respond');
```

It doesn't work when running directus directly from the git repo however, which would be handy for debugging.

**Reply by maxbritto** [reply]
Update 2022 : The Endpoint full example link was broken, here is the new link : [https://docs.directus.io/extensions/endpoints.html#example-recipes](https://docs.directus.io/extensions/endpoints.html#example-recipes)
